packages/mask/src/components/shared/WalletStatusBox/TransactionList.tsx

```diff
@@ -0,0 +1,185 @@
+    [TransactionStatusType.FAILED]: 'Failed',
+    [TransactionStatusType.CANCELLED]: 'Cancelled',
+}
+
+const getContractFunctionName = async (data: string | undefined) => {
+    if (!data) return null
+    const sig = data.slice(0, 10)
+    const name = await Services.Ethereum.getContractFunctionName(sig)
+    return name ? name.replace(/_/g, ' ') : 'Contract Interaction'
+}
+
+interface TransactionProps extends GridProps {
+    chainId: ChainId
+    transaction: RecentTransaction
+    onClear?(tx: RecentTransaction): void
+}
+const Transaction: FC<TransactionProps> = ({ chainId, transaction: tx, onClear = noop, ...rest }) => {
+    const { t } = useI18N()
+    const { Utils } = useWeb3State()
+    const { classes, theme } = useStyles()
+
+    const web3 = useWeb3()
+    const { value: targetAddress } = useAsync(async () => {
+        if (tx.receipt?.contractAddress) return tx.receipt.contractAddress
+        if (tx.payload?.params?.[0].to) return tx.payload.params[0].to as string
+        const transaction = await web3.eth.getTransaction(tx.hash)
+        return transaction.to
+    }, [web3, tx])
+    const address = (targetAddress || '').toLowerCase()
+    const txData = tx.payload?.params?.[0].data as string | undefined
+    const { value: functionName } = useAsync(async () => {
+        return getContractFunctionName(txData)
+    }, [txData])
+
+    const handleClear = useCallback(() => {
+        onClear(tx)
+    }, [onClear, tx])
+
+    const domain = useMemo(() => getContractOwnerDomain(address), [address])
+
+    const [txStatus, setTxStatus] = useState(tx.status)
+
+    useEffect(() => {
+        return WalletMessages.events.transactionStateUpdated.on((data) => {
+            switch (data.type) {
+                case TransactionStateType.CONFIRMED:
+                    if (tx.hash === data.receipt.transactionHash) {
+                        setTxStatus(TransactionStatusType.SUCCEED)
+                    }
+                    break
+                case TransactionStateType.FAILED:
+                    if (tx.hash === data.receipt?.transactionHash) {
+                        setTxStatus(TransactionStatusType.FAILED)
+                    }
+            }
+        })
+    }, [tx.hash])
+
+    return (
+        <Grid container {...rest}>
+            <Grid item className={classes.cell} textAlign="left" md={5}>
+                <Stack>
+                    <Typography className={classes.methodName} variant="body1" component="strong">
+                        {functionName}
+                    </Typography>
+                    <Typography variant="body1" color={theme.palette.text.secondary}>
+                        {format(tx.at, 'yyyy.MM.dd hh:mm')}
+                    </Typography>
+                </Stack>
+            </Grid>
+            <Grid item className={classes.cell} flexGrow={1} md={4} justifyContent="right">
+                {address && domain === address ? Utils?.formatAddress?.(address, 4) : domain || address}
```

Comment:
```suggestion
                {address && Utils.isSameAddress(domain, address) ? Utils?.formatAddress?.(address, 4) : domain || address}
```

Reply:
We should have a `Utils.formatAddressOrDomain()`.

Reply:
done.